src/flask_pyoidc/provider_configuration.py

```diff
@@ -164,22 +181,45 @@ def ensure_provider_metadata(self):
     def registered_client_metadata(self):
         return self._client_metadata
 
-    def register_client(self, redirect_uris, extra_parameters=None):
+    def register_client(self, client: Client, redirect_uri: str):
+
         if not self._client_metadata:
-            if 'registration_endpoint' not in self._provider_metadata:
+            if not self._provider_metadata['registration_endpoint']:
                 raise ValueError("Can't use dynamic client registration, provider metadata is missing "
                                  "'registration_endpoint'.")
 
             registration_request = self._client_registration_info.to_dict()
-            registration_request['redirect_uris'] = redirect_uris
-            if extra_parameters:
-                registration_request.update(extra_parameters)
-
-            resp = self.requests_session \
-                .post(self._provider_metadata['registration_endpoint'],
-                      json=registration_request,
-                      timeout=self.DEFAULT_REQUEST_TIMEOUT)
-            self._client_metadata = ClientMetadata(redirect_uris=redirect_uris, **resp.json())
+            # Check if the user has passed list of redirect_uris in ClientRegistrationInfo.
+            if 'redirect_uris' not in registration_request:
+                # If not, create it.
+                registration_request['redirect_uris'] = [redirect_uri]
+            # if it's passed, check if it's type list.
+            elif not isinstance(registration_request['redirect_uris'], list):
+                raise ValueError('redirect_uris must be a list, not str.')
+            else:
+                # Add redirect_uri into the list of redirect_uris passed by
+                # the user in ClientRegistrationInfo.
+                registration_request['redirect_uris'].append(redirect_uri)
+            post_logout_redirect_uris = []
+            # Check if post_logout_redirect_uris exists.
+            if 'post_logout_redirect_uris' in registration_request:
+                post_logout_redirect_uris = registration_request['post_logout_redirect_uris']
+                # If it exists, add this into the list of redirect_uris.
+                registration_request['redirect_uris'].extend(post_logout_redirect_uris)
+            # Remove duplicate redirect_uris from the list.
+            registration_request['redirect_uris'] = list(set(registration_request['redirect_uris']))
+
+            # Send request to register the client dynamically.
+            registration_response = client.register(
```

Comment:
This should not use `client.register` function, but keep the `self.requests_session` usage instead.
That might be configured differently, e.g. with timeouts, etc.

Reply:
I've given the reason why we should only use base package functions to interface with IdP in  #128.

Reply:
For consistency it should use the existing setup (which might be improved at a later date).

Reply:
PR to change rest of the raw requests is already complete. I'll open that one once this one is merged. It's not feasible for me to rework this because it's corresponding test cases takes unnecessary time to undo the change.

What we can do is once there are no more requested changes to do, I'll open that PR. Then you can merge both PR one after the another at the same time.

Reply:
All tests still pass if I simple revert this changed line to what it is in `main` branch: 
```python 
registration_response = client.register(
                url=self._provider_metadata['registration_endpoint'],
                **registration_request)
```

Reply:
Still this one left, please change it back for simple merging of this before fixing all of these in the next PR.

Reply:
This requires lot of rework to undo it, specially its test cases. You have to give an exception for this module. 😥

Reply:
The changes in tests were really minor, but I'll make an exception this time.
I spotted a tiny security issue that I took the liberty to fix as well - so I'll merge this now.

Thanks so much again! 💯